### PR TITLE
New Top-Down Pose Estimation Model - RTMPose

### PR DESCRIPTION
**Pull request by n-poulsen**
This pull requests adds the implementation for [RTMPose: Real-Time Multi-Person Pose Estimation](https://arxiv.org/pdf/2303.07399).

Full changes:

- New mixin for backbones with pretrained weights on [HuggingFace ](https://huggingface.co/DeepLabCut/DeepLabCut-Backbones): `HuggingFaceWeightsMixin`
- New backbone: `CSPNeXt`
- New head: `RTMCCHead`
- New target generator: `SimCCGenerator`
- New predictor: `SimCCPredictor`
- New weight initializer: `Rtmpose` 
- New `RandomBBoxTransform` augmentation for top-down models (jitters ground truth bounding boxes)
- Configs for `CSPNeXt` backbones: `cspnext_s`, `cspnext_m`, `cspnext_x`
- Configs for `RTMPose` models: `rtmpose_s`, `rtmpose_m`, `rtmpose_x`
- Updated way to scheduler configurations, to allow for the use of `SequentialLR` (to match the RTMPose training schedulers)
- Added documentation: using a `SequentialLR` scheduler for warm-up epochs
- Updated the top-down crop to function with no-square crops (and re-implemented with numpy)
  - Added `top_down_crop_size: (crop_w, crop_h)` and ` top_down_crop_margin` in the `pytorch_config.yaml` file for an easy way to control the size of the crops for top-down models.
- Bug fix: `bbox_from_keypoints` when there's more than 1 individual (bug had no impact as method was always called in a way it wouldn't fail)

DEMO

https://github.com/user-attachments/assets/173fba9c-1f82-4385-b014-add22d1ff693

### CODE REVIEW

File: deeplabcut/pose_estimation_pytorch/config/rtmpose/rtmpose_x.yaml [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+      shift_factor: 0.16
+      shift_prob: 0.3
+      scale_factor: [ 0.75, 1.25 ]
+      scale_prob: 1.0
+      p: 1.0
+    top_down_crop:
+      width: 384
+      height: 384
+method: td  # Need to add a detector
+model:  # https://github.com/open-mmlab/mmpose/blob/main/projects/rtmpose/rtmpose/body_2d_keypoint/rtmpose-x_8xb256-700e_coco-384x288.py
```
Comment (maximpavliv) [inline deeplabcut/pose_estimation_pytorch/config/rtmpose/rtmpose_x.yaml]: Has the comment been added on purpose here or is it a mistake?
What about other comments in the model configurations (`# Need to add a detector` etc)?

File: deeplabcut/pose_estimation_pytorch/data/dataset.py [hunk 2]
```diff
@@ -17,13 +17,14 @@
 import numpy as np
 from torch.utils.data import Dataset
 
+from deeplabcut.pose_estimation_pytorch.data.image import top_down_crop
 from deeplabcut.pose_estimation_pytorch.data.utils import (
-    _crop_and_pad_image_torch,
     _crop_image_keypoints,
     _extract_keypoints_and_bboxes,
```
Comment (maximpavliv) [inline deeplabcut/pose_estimation_pytorch/data/dataset.py:23]: Should we remove the underscore from these methods names (and add a backwards compatibility update line)?

File: deeplabcut/pose_estimation_pytorch/data/transforms.py [hunk 3]
```diff
@@ -557,3 +569,104 @@ def _keypoint_in_hole(self, keypoint, hole: tuple[int, int, int, int]) -> bool:
...
+
+        # Albumentations provides bounding boxes in normalized xyxy format internally
+        bboxes_xyxy = np.asarray(bboxes)
+        bboxes_extra = None
+        if bboxes_xyxy.shape[1] > 4:
+            # can't take from array - may have different dtype
+            bboxes_extra = [bbox[4:] for bbox in bboxes]
+            bboxes_xyxy = bboxes_xyxy[:, :4]
+
+        # TODO(niels): should bboxes width and height be scaled with same value?
```
Comment (maximpavliv) [inline deeplabcut/pose_estimation_pytorch/data/transforms.py]: Should we address this comment before merging?

File: deeplabcut/pose_estimation_pytorch/models/weight_init.py [hunk 4]
```diff
@@ -4,6 +4,7 @@
 
 from abc import ABC, abstractmethod
 
+import torch.nn
```
Comment (maximpavliv) [inline deeplabcut/pose_estimation_pytorch/models/weight_init.py]: Let's either remove this line or the line just below

File: tests/pose_estimation_pytorch/data/test_transforms.py [hunk 5]
```diff
@@ -101,3 +103,213 @@ def test_coarse_dropout():
...
+
+    gen = np.random.default_rng(seed=0)
+
+    # generate 100 bboxes
+    num_bboxes = 100
+    bboxes = np.zeros((num_bboxes, 4))
+    # sample x, y in the images
+    bboxes[:, :2] = image_wh * gen.random((num_bboxes, 2))
+    # sample w, h with the space remaining
+    bboxes[:, 2:] = (image_wh - bboxes[:, :2]) * gen.random((num_bboxes, 2))
```
Comment (maximpavliv) [inline tests/pose_estimation_pytorch/data/test_transforms.py]: The bboxes generation is duplicated across tests, let's move this out into a separate method?

**Review by maximpavliv — CHANGES_REQUESTED** [review]

File: deeplabcut/pose_estimation_pytorch/data/dataset.py (hunk 2 above)
Comment (n-poulsen) [reply deeplabcut/pose_estimation_pytorch/data/dataset.py:23]: we can and probably should, in a bigger refactor of the dataset class; but this shouldn't be done in this pull request as it concerns other parts of the code (and not RTMPose implementation)

File: deeplabcut/pose_estimation_pytorch/config/rtmpose/rtmpose_x.yaml (hunk 1 above)
Comment (n-poulsen) [reply deeplabcut/pose_estimation_pytorch/config/rtmpose/rtmpose_x.yaml]: The github link is forgotten, I'll remove it; the `detector` comment is added for all top-down models before a detector is added

File: deeplabcut/pose_estimation_pytorch/data/transforms.py (hunk 3 above)
Comment (n-poulsen) [reply deeplabcut/pose_estimation_pytorch/data/transforms.py]: Yes I tested that but forgot the todo - I'll take care of it

**Review by maximpavliv — APPROVED** [review]
Great!!
